watch.go

```diff
@@ -103,48 +121,54 @@ func (fb *Firebase) Watch(notifications chan Event) error {
 			// 		event: put
 			// 		data: {"path":"/","data":{"foo":"bar"}}
 
-			var evt []byte
-			var dat []byte
-			isPrefix := true
-			var result []byte
-
-			// For possible results larger than 64 * 1024 bytes (MaxTokenSize)
-			// we need bufio#ReadLine()
-			// 1. step: scan for the 'event:' part. ReadLine() oes not return the \n
-			// so we have to add it to our result buffer.
-			evt, isPrefix, scanErr = scanner.ReadLine()
+			payload := make([]string, 2)
+			scanErr = func(payload []string) error {
+				lineCount := 0
+				for {
+					line, err := scanner.ReadString('\n')
+					if err != nil {
+						return err
+					}
+					line = strings.Trim(line, " \r\n")
+					if len(line) == 0 { // empty line
+						if lineCount == len(payload) {
+							return nil // everything OK
+						} else {
+							return errors.New("Bad formated body")
+						}
+					}
+					if lineCount < len(payload) {
+						payload[lineCount] = line
+						lineCount++
+					}
+				}
+			}(payload)
 			if scanErr != nil {
 				break scanning
 			}
-			result = append(result, evt...)
-			result = append(result, '\n')
-
-			// 2. step: scan for the 'data:' part. Firebase returns just one 'data:'
-			// part, but the value can be very large. If we exceed a certain length
-			// isPrefix will be true until all data is read.
-			for {
-				dat, isPrefix, scanErr = scanner.ReadLine()
-				if scanErr != nil {
-					break scanning
-				}
-				result = append(result, dat...)
-				if !isPrefix {
-					break
-				}
-			}
-			// Again we add the \n
-			result = append(result, '\n')
-			_, _, scanErr = scanner.ReadLine()
-			if scanErr != nil {
+
+			var eventType string
+			if strings.HasPrefix(payload[0], "event:") {
```

Comment:
No problem. I will do this.

█  Jacques

On 31 July 2016 at 20:56, Steven Berlanga <notifications@github.com> wrote:

> In watch.go
> <https://github.com/zabawaba99/firego/pull/50#discussion_r72908889>:
>
> > -				dat, isPrefix, scanErr = scanner.ReadLine()
> > -				if scanErr != nil {
> > -					break scanning
> > -				}
> > -				result = append(result, dat...)
> > -				if !isPrefix {
> > -					break
> > -				}
> > -			}
> > -			// Again we add the \n
> > -			result = append(result, '\n')
> > -			_, _, scanErr = scanner.ReadLine()
> > -			if scanErr != nil {
> > +
> > +			var eventType string
> > +			if strings.HasPrefix(payload[0], "event:") {
>
> Can you handle the negative case first?
>
> if !strings.HasPrefix(payload[0], "event:") {
>     scanErr = errors.New("First line does not start with event:")
>     break scanning
> }
> // do the string trimming
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zabawaba99/firego/pull/50/files/91637d68c23bcd6761930b23882b18621b759365#r72908889>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAB2iAUufoxcam9Yu4VVm3r_qNYBGXVkks5qbO_HgaJpZM4JYHTS>
> .
>